sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,357 @@
+    val newOutputMap = AttributeMap(newExpressions.map { ne =>
+      val mapped = mapAttributes(ne, outputMap)
+      val withoutAlias = mapped match {
+        case Alias(child, _) => child
+        case e => e
+      }
+      ne.toAttribute -> mergedExpressions.find {
+        case Alias(child, _) => child semanticEquals withoutAlias
+        case e => e semanticEquals withoutAlias
+      }.getOrElse {
+        mergedExpressions += mapped
+        mapped
+      }.toAttribute
+    })
+    (mergedExpressions.toSeq, newOutputMap)
+  }
+
+  // Only allow aggregates of the same implementation because merging different implementations
+  // could cause performance regression.
+  private def supportedAggregateMerge(newPlan: Aggregate, cachedPlan: Aggregate) = {
+    val newPlanAggregateExpressions = newPlan.aggregateExpressions.flatMap(_.collect {
+      case a: AggregateExpression => a
+    })
+    val cachedPlanAggregateExpressions = cachedPlan.aggregateExpressions.flatMap(_.collect {
+      case a: AggregateExpression => a
+    })
+    val newPlanSupportsHashAggregate = Aggregate.supportsHashAggregate(
+      newPlanAggregateExpressions.flatMap(_.aggregateFunction.aggBufferAttributes))
+    val cachedPlanSupportsHashAggregate = Aggregate.supportsHashAggregate(
+      cachedPlanAggregateExpressions.flatMap(_.aggregateFunction.aggBufferAttributes))
+    newPlanSupportsHashAggregate && cachedPlanSupportsHashAggregate ||
+      !newPlanSupportsHashAggregate && !cachedPlanSupportsHashAggregate && {
+        val newPlanSupportsObjectHashAggregate =
+          Aggregate.supportsObjectHashAggregate(newPlanAggregateExpressions)
+        val cachedPlanSupportsObjectHashAggregate =
+          Aggregate.supportsObjectHashAggregate(cachedPlanAggregateExpressions)
+        newPlanSupportsObjectHashAggregate && cachedPlanSupportsObjectHashAggregate ||
+          !newPlanSupportsObjectHashAggregate && !cachedPlanSupportsObjectHashAggregate
+      }
+  }
+
+  // Second traversal replaces `ScalarSubqueryReference`s to either
+  // `GetStructField(ScalarSubquery(CTERelationRef to the merged plan)` if the plan is merged from
+  // multiple subqueries or `ScalarSubquery(original plan)` if it isn't.
+  private def removeReferences(
+      plan: LogicalPlan,
+      cache: ListBuffer[Header]): (LogicalPlan, Seq[CTERelationDef]) = {
+    val subqueryCTEs = mutable.Map.empty[Int, CTERelationDef]
+    val newPlan =
+      plan.transformAllExpressionsWithPruning(_.containsAnyPattern(SCALAR_SUBQUERY_REFERENCE)) {
+        case ssr: ScalarSubqueryReference =>
+          val header = cache(ssr.subqueryIndex)
+          if (header.merged) {
+            val subqueryCTE = subqueryCTEs.getOrElseUpdate(ssr.subqueryIndex,
+              CTERelationDef(createProject(header.elements, header.plan)))
+            GetStructField(
+              ScalarSubquery(
+                CTERelationRef(subqueryCTE.id, _resolved = true, subqueryCTE.output,
```

Comment:
If we can force the CTE reuse, then I think a better plan here is
```
ScalaSubquery(
  Project(attr_at_header_index,
    CTERelationRef(...))
)
```

Reply:
There is always only one output attribute (`mergedValue`) in a merged subquery CTE e.g.: https://github.com/apache/spark/blob/96ed6fd72ca262e88b3c0469ac57f994b1d1c905/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala#L74-L84 so I'm not sure it makes sense to project `mergedValue` from `CTERelationRef`.

The `GetStructField`s wrap the `ScalarSubquery` expressions so the subquery plans are identical and can be reused in the physical plan.